File: backend/app/schemas/malla.py
from enum import Enum
from typing import List, Optional
from pydantic import BaseModel


class StatusCurso(str, Enum):
    COMPLETED = "completed"
    IN_PROGRESS = "in_progress"
    AVAILABLE = "available"
    LOCKED = "locked"


class PrerrequisitoInfo(BaseModel):
    id: str
    code: str
    name: str
    completado: bool


class CourseDetail(BaseModel):
    id: str
    code: str
    name: str
    credits: int
    status: StatusCurso
    description: Optional[str] = None
    # Binario, derivado del estado. El avance fino dentro de un curso (por
    # unidades) es RF-11 y vive en el detalle de curso, no en la malla.
    progreso: int = 0
    # Datos del historial del estudiante. Solo vienen si tiene registro en
    # progreso_cursos; un curso que nunca llevó los deja en None.
    nota: Optional[float] = None
    fecha_completado: Optional[str] = None
    prerequisitos: List[PrerrequisitoInfo] = []
    prerequisitos_cumplidos: bool = True


class ResumenCiclo(BaseModel):
    """Conteo de estados del ciclo, para no recalcularlo en cada pantalla."""

    total: int = 0
    aprobados: int = 0
    en_curso: int = 0
    disponibles: int = 0
    bloqueados: int = 0
    creditos_aprobados: int = 0


class CicloDetail(BaseModel):
    # Etiqueta lista para mostrar ("Ciclo 3").
    ciclo: str
    # El número en crudo. Sin este campo el frontend tenía que hacer
    # ciclo.split(" ")[1] para recuperarlo, y eso se rompe apenas cambie el
    # texto de la etiqueta o se traduzca.
    ciclo_num: int
    credits: int
    resumen: ResumenCiclo = ResumenCiclo()
    courses: List[CourseDetail]

File: frontend/types/malla.ts
// Malla curricular: tipos espejo del contrato del backend (GET /api/malla).
// Fuente de verdad: backend/app/schemas/malla.py. No definir aquí campos que
// el backend no envía; si cambia el contrato, esto debe fallar en build.

export type StatusCurso = "completed" | "in_progress" | "available" | "locked"

export interface PrerrequisitoInfo {
  id: string
  code: string
  name: string
  completado: boolean
}

export interface CourseDetail {
  id: string
  code: string
  name: string
  credits: number
  status: StatusCurso
  description?: string | null
  // Binario, derivado del estado. El avance fino por unidades es RF-11.
  progreso: number
  // Solo vienen si el estudiante tiene registro en progreso_cursos.
  nota?: number | null
  fecha_completado?: string | null
  // Prerrequisitos DIRECTOS (RF-06): de qué cursos cuelga este.
  prerequisitos: PrerrequisitoInfo[]
  // De toda la cadena, los que aún no aprueba (explican el candado).
  prerequisitos_faltantes: PrerrequisitoInfo[]
  prerequisitos_cumplidos: boolean
}

export interface ResumenCiclo {
  total: number
  aprobados: number
  en_curso: number
  disponibles: number
  bloqueados: number
  creditos_aprobados: number
}

export interface CicloDetail {
  // Etiqueta lista para mostrar ("Ciclo 3").
  ciclo: string
  // Número en crudo; el frontend no debe parsear el string.
  ciclo_num: number
  credits: number
  resumen: ResumenCiclo
  courses: CourseDetail[]
}

// Respuesta de GET /api/malla/avance (RF-07). La cifra oficial del plan.
export interface AvanceCarrera {
  carrera_id?: number
  malla_id?: number
  porcentaje_avance: number
  creditos_aprobados: number
  creditos_en_curso: number
  creditos_totales: number
  creditos_restantes: number
  cursos_aprobados: number
  cursos_en_curso: number
  cursos_totales: number
}
